### ISSUE TITLE
Move inputsets to pymatgen?

### ISSUE THREAD

**Issue by computron** [materialsproject/atomate2#486]
There are a bunch of new VaspInputSets in atomate2:

https://github.com/materialsproject/atomate2/tree/main/src/atomate2/vasp/sets

Shyue Ping has suggested to move them into pymatgen. I wanted to bring up a thread to discuss pros and cons about leaving in atomate2 vs leaving in pymatgen. In particular, if we keep them separate, we need to have clarity on what is the "official recommended input sets" of Materials Project so that people don't get confused about what to use if they want to do the same thing as MP, or want the "standard recommended" input set.

**Comment by janosh** [comment]
Funny, was just about to open this same issue after discussing with @arosen93. See related https://github.com/Quantum-Accelerators/quacc/issues/718.

**Comment by arosen93** [comment]
I think having them in one place is important, which I suppose means my vote is for Pymatgen. I'm having trouble thinking of many pros for keeping them in atomate2, aside from it being a more natural place for them. But I think that is outweighed by the cons associated with having the "same" (which may not ultimately be the same...) data in two places.

**Comment by janosh** [comment]
> As a side note, it might be worth making it clearer in the MP docs that the "one true source" of input set data is in Pymatgen because before joining the group here I had no idea those modules even existed.

Were you thinking of the `pymatgen` docs? Happy to add a paragraph.

**Comment by arosen93** [comment]
@janosh: I was thinking in the MP Documentation, e.g. [here](https://docs.materialsproject.org/methodology/materials-methodology/calculation-details/gga+u-calculations/pseudopotentials) and [here](https://docs.materialsproject.org/methodology/materials-methodology/calculation-details/r2scan-calculations/pseudopotentials). An easy add.

**Comment by arosen93** [comment]
One con for having the sets live only in Pymatgen is that users of Atomate2 might not read the release notes of Pymatgen. Then if they update Pymatgen (either intentionally or indirectly as part of some other package update), they might no longer have internally compatible input settings for their calculations. If they lived in Atomate2, then this would likely be easier to spot.

**Comment by janosh** [comment]
This is a responsibility of the package maintainers, i.e. us, to highlight in the `atomate2` release notes whenever we update dependencies if the changes include anything that materially affects users (like an input set change in `pymatgen`).

**Comment by arosen93** [comment]
@janosh, @esosteric-ephermera: If we _do_ ultimately decide to keep input sets in pymatgen, this is just a reminder to confirm that your very reasonable changes in #362 get carried over (e.g. setting some reasonable default for KSPACING, SIGMA, ISMEAR and changing LELF). If we do keep two separate input sets, it would likely be worth updating the Pymatgen ones too because the r2SCAN input set is basically unusable without these changes then.

**Comment by esoteric-ephemera** [comment]
Agreed that moving these to pymatgen likely makes the most sense. Would also help if we renamed the sets something like "CurrentMPGGASet" to distinguish from the legacy ones that are only compatible with older releases of MP

**Comment by JaGeo** [comment]
Hi all, would it be possible to do a atomate2 meeting including the European developers (at a Germany and UK friendly time)? I have the feeling that a lot of changes are currently discussed.

**Comment by utf** [comment]
I think the plan was always to move the input sets to pymatgen at some point. A few things to think about:
1. We currently maintain two classes of input sets. The MP compatible ones (being developed in #352) and the default atomate2 input sets that use PBEsol rather than PBE. Do we want to move all input sets across or just the MP compatible ones?
2. The input set sets in atomate2 are fundamentally different to the ones in pymatgen. @rkingsbury developed a new input set generator interface (see https://github.com/materialsproject/pymatgen/pull/2211) which has a different API to the existing `DictSet` based VASP input sets in pymatgen. Do we plan to i) replace the old pymatgen input sets (not recommended since this would break a lot of code, e.g. atomate1) or ii) have these new input sets live alongside the old ones? If we go for option ii, we still have the issue of maintaining two generations of input sets.

**Comment by janosh** [comment]
> Do we want to move all input sets across or just the MP compatible ones?

Re 1. I don't have strong views but leaning towards just the MP-compatible ones. I think @shyuep's main concern is duplication and the potential for subtle historical drift. Things that are only in atomate2 don't have that. On the other hand, I'm not against having a single location for all inputs sets, i.e. moving both.

> If we go for option ii, we still have the issue of maintaining two generations of input sets.

Re 2. what issues are you foreseeing? Confusion on users part? Or the maintenance burden in itself? Both seem manageable to me.

@JaGeo Yes, I suggested that last week. @computron and @Zhuoying are in the process of finding a new time for the atomate2 meeting 1 or 2 hours earlier.

**Comment by shyuep** [comment]
@utf That is why I have always been against this. Why is there a NEW vasp input set interface in atomate2 coded by @rkingsbury  that now is in conflict with the pymatgen one? Had all these been done in pymatgen from the start, we would have resolved this earlier rather than waiting until now where the entire atomate2 works on one system and pymatgen works on another. Now we either have to choose to live with both or someone has to undertake the massive work to merge/modify one side to use the other. In short, drift has now already occurred.

Bottomline: Don't do work in another repo assuming you will someday "move" it to somewhere else. It never ever works out.

In any case, I can only share my experience and not actually control what you guys want to do. I am not having two ways of defining input sets in pymatgen. If you want to replace old Pymatgen input sets with the new ones, you figure out how to do that in a backward-compatible manner (breaking backwards compatibility is **not** an option). If you want to discard the new input sets and integrate some improvements to the pymatgen ones, go ahead. If you want to continue on the path of atomate2 having its own input set definition forever, so be it.

**Comment by shyuep** [comment]
@arosen93 The reason why the input sets should be in pymatgen is because people often do MP compatible calculations without using atomate2 at all (I myself do this extensively and certainly my group members do). Why is there a need to insist that they must have atomate2 to do a calculation to get an Ehull using MPRester? 

This is more or less akin to a situation where the atomate2 devs suddenly decide that pymatgen.io.vasp doesn't meet all their needs, and instead of modifying pymatgen.io.vasp to meet their needs, they proceed to implement their own versions of Incar, Potcar, Poscar and Potcar with a completely different attributes and methods API, and a year down the road say "oh no, maybe we should make two versions of Incars, two Poscars, etc.".

**Comment by arosen93** [comment]
> @arosen93 The reason why the input sets should be in pymatgen is because people often do MP compatible calculations without using atomate2 at all (I myself do this extensively and certainly my group members do). Why is there a need to insist that they must have atomate2 to do a calculation to get an Ehull using MPRester?

Just noting here that I agree with you. I don't think I suggested otherwise, but if I did, that was poor communication on my part. 😉

**Comment by shyuep** [comment]
@utf If you read my comment (from 2021!!), it was clear that the VaspInputGenerator should have been done within the pymatgen context. Not in a separate code base. If you are expecting that the pymatgen maintainers are going to follow the developments of a separate code base (on top of dealing with all the day-to-day issues), that's not reasonable. 

As of now, given that I have not followed whatever VaspInputGenerator thing that atomate2 supposedly works on, I have not actually seen the evidence. Note the sentence "Once we are satisifed that the generators are a complete replacement for VaspInputSets and have **demonstratable benefits**, we will deprecate VaspInputSets." Feel free to make the case and then we decide if VaspInputSets will be deprecated or not (with major consequences of the rest of the community).

**Comment by shyuep** [comment]
@arosen93 I know you were agreeing with me. :-) I am just explaining the reason.

**Comment by utf** [comment]
@shyuep Yes, I acknowledged that the input sets should have been moved across earlier. However, let's say the original timeframe was for the input sets to live in parallel for 1 year or more. I don't see any reason why it makes a difference if we start that process now versus in March 2022 (when that PR was merged). 

One proposal is that we could keep two interfaces but have a single source of logic for generating the input sets. That would involve:
1. Adding the generators to somewhere in pymatgen.
2. Updating the old input sets to use the generators internally while still exposing their original API.

Obviously, the PR that contains that change would detail any and all modifications to the input settings. In general, we have tried to make those changes as minimal as possible, so I don't imagine this will be much.

**Comment by utf** [comment]
The reasons for the API change were outlined in the PR already mentioned. It was discussed in detail, with specific examples given of why this was necessary. Ultimately, whether you think there are demonstrable benefits comes down to the points made in that PR. Regardless, the abstract interface already lives in pymatgen and is already implemented for a range of simulation packages in pymatgen itself. 

Ultimately, there is no going back for atomate2 (it is critical that we can initialise an input set generator without a structure).  If we want atomate1 to keep running, then we need the old input set interface otherwise atomate1 will have to be refactored and that will be a waste of everyone's time. 

Having two interfaces where one is deprecated later on is not unpythonic. If we can make sure that the actual logic for generating the input settings is universal then there will be no drift between the settings. It will cause the least amount of hassle for developers and users.

**Comment by JaGeo** [comment]
> @JaGeo Yes, I suggested that last week. @computron and @Zhuoying are in the process of finding a new time for the atomate2 meeting 1 or 2 hours earlier.
@janosh great, thank you!

**Comment by shyuep** [comment]
@utf I did read the PR (again, twice). That PR merely introduced the abstract classes. There are *claimed* benefits but there are no demonstrated benefits. As I already wrote, I wanted to see demonstrated benefits. That has never been the case. If you want to point out to me where in that very long chain that I agreed the deprecate the old VaspInputSets, by all means. I can see no such unconditional agreement anywhere.

Right now, all you have been doing is to merely make very general statements such as "it is critical that" or "there is no going back for atomate2" and somehow expect the pymatgen side to accommodate all these changes at a huge hassle based merely on those statements. Why is it "no going back" for atomate2, but somehow pymatgen is expected to be "going back" on the existing input sets?

Anyway, it is up to you whether you want to take the trouble to provide the information or not. If you do, I will deprecate the old VaspInputSets. If you prefer not to, we will just live with input sets living in two universes.

**Comment by matthewkuner** [comment]
I've recently been working on creating a validator to accept external calcs into the MPDB. From the point of view of validating calculations (especially external calcs), having one consistent "ground truth" input set is absolutely essential. Two input sets would make crowd-sourcing the MPDB terribly confusing

**Comment by shyuep** [comment]
@matthewkuner As an aside, I would say that accepting external calculations into MPDB as-is is not a good practice. One of the main competitive value of MP is that the DB is relatively clean - the calculations were done using a consistent set of parameters. Even then, MPDB has consistency problems with calculations from a long time ago being different enough from recent calculations.

Instead of just accepting the calculations, a better way would be to take those calculations, re-relax them using MP and redo static calculations. If the calculations were properly done in the first place, the relaxation would be 1-2 steps. It will take a negligible computational cost (for reference, @janosh did 60k SCAN statics in a few days). If they are taking longer, they are not properly done anyway and you should be throwing those calculations out.

**Comment by shyuep** [comment]
@matthewkuner I should also point out that there is no way to properly validate anyway. Say I have done a proper MP input set calculation but I really really want to corrupt the MP data. I simply do a sed to mess the energy values and MP will be screwed. I can trojan horse an ultrastable compound into MP and all the phase diagrams in that system will be nonsense.

**Comment by matthewkuner** [comment]
@shyuep I guess it would depend on how much we are willing to trust external users. Perhaps people would have to apply/get approved to send in their calcs. 

And either way, having one "ground truth" input set for validating our internal calcs is still vital

**Comment by shyuep** [comment]
@matthewkuner I agree one ground truth for input sets is important. That wasn't the point of my comment.

I trust 99% of people. But this is a situation where even a single bad player can have a massive impact. The cost of minimizing that risk is so little. Anyway, that's my input. BTW, if MP trusts users so much, maybe lift all restrictions on the MP-API? ;-)

**Comment by computron** [comment]
All - this is getting off-track. Note that in terms of input sets, we are not discussing the issue of having multiple different input sets here. We were only discussing (and mainly rejecting) whether the same settings could be repeated in multiple repos, or twice within the same repo, as an implementation detail. This is different than whether MP supports or uses different settings.

If there is a discussion regarding how and why MP needs to validate input sets, or if MP should be re-calculating, etc. that goes somewhere else than this atomate2 issue.

**Comment by computron** [comment]
We discussed this issue in a meeting on 9/1 (see attached slides on discussion points)

[vis_discussion.pptx](https://github.com/materialsproject/atomate2/files/12500234/vis_discussion.pptx)

The decision from the meeting was that the existing pymatgen input sets would be modified to allow for the use cases defined in the slides. It was left a bit unclear at the end of the meeting whether this would happen by working from the existing pymatgen input sets and changing those, or working from the new atomate2 input sets and moving that into pymatgen. However, regardless of the approach, **backward compatibility for previous users of pymatgen was going to be maintained.**

The approach discussed at the meeting was:
1. A user will be allowed to set structure=None when constructing a VaspInputSet (VIS). Please note that some VIS require other information like band_gap to function properly, so it's not just structure we are talking about.
2. VIS will provide a set_structure() method to allow for delayed setting of the structure. Again, there may need to be delayed input of other parameters like band_gap for some VIS.
3. VIS will provide a new get_input_set(structure) method that calls set_structure() and then generates the inputs. Note that both myself and @matthewkuner expressed some reservations about whether this was necessary, and whether it may cause extra confusion.

There were also some secondary improvements discussed and suggested by @utf, like:
1. Making it easy to define kpoint modes supported by VIS
2. Cleaning up the from_prev() logic and moving all logic to the same place
3. Removing multiple layers of inheritance of VIS to make debugging easier and surprises less frequent (this one had some pros / cons associated with it)

To support these secondary improvements, @utf suggested it may be cleaner to move the atomate2 VIS implementations (technically "Makers") into pymatgen instead of making modifications to the PMG VIS. Most people seemed to agree that as long as backward compatibility was retained and the final solution looked like what was discussed, the approach taken to get to that solution was not all that important to them.

Overall, this would remove the split of VIS between pymatgen and atomate2, and should lead to better outcomes for all involved after the initial messy code refactoring process.

As for who does the refactoring, I believe @utf was volunteering to look at it and @shyuep was also thinking to take a crack at it. However, no real execution plan was made at the meeting so this should probably be coordinated between @utf and @shyuep to avoid issues. I'm happy to work with them on this if they need it.

**Comment by shyuep** [comment]
The basic pymatgen input changes are done: See https://github.com/materialsproject/pymatgen/commit/fabdd276497493f87f72348a3573bbe8a6c38f26 

Key changes:
- structure = None by default.
- Instead of set_structure, I used as property setter. All structure manipulations (change of setting etc.) are all performed in the setter method. This is the pythonic way.
- If vis.incar, etc. are called without a valid structure, then a RuntimeError is raised. 
- The init now contains minimal processing (mainly to deal with things like user_incar or user_potcar settings). But these can be moved to the respective methods if desired.
- get_vasp_input also now takes an optional structure.
- A very basic test is added for MPRelaxSet with and without structure.

It was relatively painless. All tests pass as is.

There are a few input sets that still use structure for processing in the init method. These have not been modified in any way. I suspect fixing these will require not too much effort.

For the rest, I will leave it to @utf to merge the existing atomate2 sets into pymatgen (feel free to override any of the code with your improvements). However, all tests should pass mostly as is, i.e., backwards compatibility is maintained.

**Comment by janosh** [comment]
That was fast. Nice work @shyuep! I also prefer the `@setter` approach. Re remaining `InputSets` using structure in their `__init__`, I think those can easily be accommodated by overriding the `DictSet` structure `@setter`.

**Comment by JaGeo** [comment]
With regard to inheritance, I just wanted to mention that many input sets inherit from `MPRelaxSet` (e.g., `LobsterSet`). Thus, it might not be completely straight-forward to remove some.of these layers.

**Comment by shyuep** [comment]
@JaGeo Yes. But I want to note that the YAML files also can do inheritance. We can make it such that the YAML files are properly separated in terms of parameters and do inheritance there. MPRelaxSet does not have any code logic whatsoever (it is all in DictSet). The only thing that MPRelaxSet does is to initialize from the yaml paramters.

**Comment by shyuep** [comment]
@JaGeo I just pushed a change that makes Lobster inherit from DictSet. It required only two small modifications and all tests pass. I believe switching to this would be relatively painless.

**Comment by JaGeo** [comment]
@shyuep I see! Thank you!

**Comment by computron** [comment]
Thanks all for your help with this issue. To continue with some atomate2 PRs we are hoping to get these input sets in their "final" form. To that end I need 2 pieces of information:

@shyuep  - I see that you added the structure addition later, however sometimes other information than structure is needed later on. For example, one of the input sets requires a band gap which cannot be known ahead of time, and other input sets in the future may also require things other than structure. This could be accommodated by changing ``set_structure`` to ``set_data`` or something generic, but wanted to ping on this as it's needed to continue.

@utf - I know you had plans to redo the base classes in a more fundamental way, but not sure if that was coordinated, in progress, etc. I wanted to check in on the status of this, particularly since @janosh would like to merge some workflows and wants some clarity on the final format of the input sets.

Hopefully it can be resolved here, otherwise I will schedule another meeting.

**Comment by shyuep** [comment]
I prefer to leave the structure setter as is. Nothing is preventing someone from writing:

```python
vis.structure = new_structure
vis.band_gap = 1.2
```

I would also note that I am unclear what is meant by "requiring a bandgap". If we are merely saying whether something is a metal or not, I think a simple setting a `is_metal` parameter would work better rather than giving a band gap value.

**Comment by computron** [comment]
>> Nothing is preventing someone from writing:

Untrue, take a look at MPScanRelaxSet. The logic for the band gap is done in the constructor, this won't be called if you set the gap later. 

The MITNEBSet is another example where you need to set not a single structure but a list of structures, and processing is done in the constructor so you can't just set the data later.

I haven't checked all the input sets rigorously for places where there may be problems, but there are certainly going to be issues. It also needs to be enforced that no logic is done in the constructor if it is acceptable behavior to change parameters later, or at least to define setters for any logic-required parameters that both set the parameter and do the logic. The previous input sets would need to look for this. Some of the more common logic is just checking if the parameter value is valid, but again this check would be skipped without real setters.

As to whether the parameter should be a numerical band gap or is_metal parameter, that is minor and I don't think it should distract from the core issue. My guess is that numerical band gap allows for more fine-tuned adjustment algorithms of k-points later down the line if we want while keeping the same plumbing intact. But anyway this is a separate issue to discuss as to what's the best parameter, band_gap or is_metal.

**Comment by shyuep** [comment]
My point is that you can move the logic for band gap setting to somewhere else. Each setter should be specific in what it modifies.

**Comment by computron** [comment]
Ok agreed, in that case someone would need to go through the existing code and modify any constructor logic (including parameter validation or internal parameter calculation) to a setter method.

It would also be nice to somehow communicate to newcomers that this is the way so that new inputsets don't have this issue of logic in the constructors.

**Comment by rkingsbury** [comment]
> It would also be nice to somehow communicate to newcomers that this is the way so that new inputsets don't have this issue of logic in the constructors.

I think that issue should be bypassed through the use of `InputGenerator`, correct? The setter logic goes there so that the returned `InputSet` does not have to be modified.

**Comment by shyuep** [comment]
I don't think it is important to inherit InputGenerator. Of course the setters should be based on the input set logic. That's what class inheritance is for.

**Comment by rkingsbury** [comment]
> I don't think `InputGenerator` is inherited by any of the `VaspInputSets`.

No it's not inherited, but the design intent (as I understood it) when we defined the new InputSet API was that all the construction logic would happen in the Generators, which then return `InputSet`. Something else may need to be done for backward compatibility on the existing sets, but going forward I think the `InputGenerator` would be the place to, for example, compute `KSPACING`.

**Comment by shyuep** [comment]
The construction logic is in the VaspInputSets. We are not renaming them to be generators. We are also not renaming the current VaspInput to InputSet. All names are kept as-is.

I don't really want to spend time rehashing all the things we already decided in the last meeting. Let me state it clearly once and for all.

1. The existing VaspInputSets now support structure-less init. All logic is in the highest level DictSet, with subclass InputSets modifying logic where necessary.
2. What you call the InputSet already was implemented in VaspInput, except for maybe a few small functions.
3. All that remains to be done is for Alex to add whatever new features/logic he has in the atomate2 input sets to the pymatgen input sets, remove the atomate2 input sets, and refactor atomate2 to use the pymatgen VaspInputSets.

The whole point why I went through all this is to satisfy the complaints that the VaspInputSets cannot be initialized in a structure-less manner (I am still doubtful of the benefits, but whatever). The concession on the atomate2 side is that there will not be any renaming to "Generators". This is a pure name change with zero practical effect except to create a gazillion headaches for pymatgen users.

**Comment by computron** [comment]
Hi all,

Let me mediate a bit here. First, I think the current solution as it stands is already workable for atomate2 without introducing any backward-incompatible changes (hurrah!). However, there are still a couple of deficiencies with the solution:

1. With the new solution, it is important for developers to *not* put any logic in the constructor but rather to define setters if they want their VIS to support delayed setting of parameters. Some of the existing VIS already violate this and need to be refactored to support delayed setting of parameters (add this the list of todos), and it is difficult to guarantee that future VIS will not break this without being on alert. It will also not cause an immediate error if the rule **is** broken upon runtime, so bugs from this could be subtle and delayed.
2. We haven't discussed the serialization yet and I haven't looked into it, but it's possible the serialization will be heavyweight for atomate2 use. This is mostly not a big deal, can be solved easily, and is mostly a thing about workflows.

I believe @utf wanted to propose another layer above VIS that helped sort some of these issues, **while also retaining full backward compatibility as in the existing solution**. But waiting for @utf as to whether this is still the plan or whether we want to proceed with what we have now.

**Comment by shyuep** [comment]
Can I have an update on this? I personally took the effort to address pymatgen's side of the bargain. I would prefer we actually have a hard deadline to make this happen?

**Comment by computron** [comment]
@utf @shyuep I think we should figure this out at the next atomate2 meeting which is on Friday Oct 27. 

@utf will you attend?

**Comment by shyuep** [comment]
I am not attending atomate meetings since I don't work on it per se. But I want to be clear that I held up my end of the bargain within a few days of our last meeting on this topic. I will put a hold on all requests made to pymatgen from atomate2 until I see this issue is completely settled. If it is never settled, so be it.

**Comment by computron** [comment]
To update post-meeting, @utf was there and has been working on some changes to the PMG input sets. He's also been working through the existing list of input sets to make sure the changes accommodate all the various things done in various input sets, including some inconsistencies in the way things are handled.

@utf plans to soon submit a WIP pull request on the changes so far. To clarify, the PR will be intended to seed discussion on the work so far rather than a specific plan to merge.

**Comment by computron** [comment]
Closing this issue as discussion is going to move to:

https://github.com/materialsproject/pymatgen/pull/3484
